lib/pbench/server/cache_manager.py

```diff
@@ -429,33 +508,45 @@ def get_info(self, path: Path) -> JSONOBJECT:
         return fd_info
 
     @staticmethod
-    def extract(tarball_path: Path, path: Path) -> Optional[IO[bytes]]:
-        """Returns the file stream which yields the contents of
-        a file at the specified path in the Tarball
+    def extract(tarball_path: Path, path: Path) -> Inventory:
+        """Returns a file stream for a file within a tarball
 
         Args:
             tarball_path: absolute path of the tarball
             path: relative path within the tarball
 
+        Returns:
+            An inventory object that mimics an IO[bytes] object while also
+            maintaining a reference to the tarfile TarFile object to be
+            closed later.
+
         Raise:
-            BadDirpath on failure extracting the file from tarball
+            TarballNotFound on failure opening the tarball
+            CacheExtractBadPath if the target cannot be extracted
         """
         try:
-            return tarfile.open(tarball_path, "r:*").extractfile(path)
+            tar = tarfile.open(tarball_path, "r:*")
         except Exception as exc:
-            raise BadDirpath(
-                f"A problem occurred processing {str(path)!r} from {str(tarball_path)!r}: {exc}"
-            )
+            raise TarballNotFound(str(tarball_path)) from exc
+        try:
+            stream = tar.extractfile(str(path))
+        except Exception as exc:
+            raise CacheExtractBadPath(tarball_path, path) from exc
+        else:
+            if not stream:
+                raise CacheExtractBadPath(tarball_path, path)
+            return Inventory(stream, tar)
 
-    def filestream(self, path: str) -> Optional[dict]:
-        """Returns a dictionary containing information about the target
-        file and file stream
+    def filestream(self, path: str) -> Optional[JSONOBJECT]:
+        """Access the file stream of a tarball member file.
 
         Args:
             path: relative path within the tarball of a file
 
-        Raises:
-            TarballUnpackError on failure to extract the named path
+        Returns:
+            An inventory object that mimics an IO[bytes] object while also
+            maintaining a reference to the tarfile TarFile object to be
+            closed later.
 
         Returns:
             Dictionary with file info and file stream
```

Comment:
We probably don't need the old `Raises` section, but we probably also don't need _two_ `Returns` sections (particularly since the new one doesn't seem to match the function's return type hint).  🙂

Reply:
Huh: wonder how *that* got there? I did wrestle this code through a succession of forms before it decided to settle down, though ...